File: lib/mockAgent.ts
// CheckMate — mock agent.
// Produces a correctly-shaped /api/execute payload with a realistic steps[] trace.
// No real LLM / OCR yet. Module names match the deck architecture (fixed pipeline):
//   Reader (OCR) -> Retriever (rubric match) -> Grader (score+feedback) -> Reflector (conditional)
// Reflection runs only when a grade is shaky and picks one action:
//   APPROVE · REVISE (<=1) · RETRY_OCR (<=1) · ESCALATE (send to teacher).

export type Step = {
  module: string;
  pattern?: string;
  prompt: { system_prompt: string; user_prompt: string };
  response: unknown;
};

export type QuestionResult = {
  id: string;
  title: string;
  score: number;
  max: number;
  status: "ok" | "partial" | "escalate";
  mark: string; // short margin stamp, e.g. "✓" / "3/5"
  feedback: string;
};

export type ExecuteResult = {
  status: "ok" | "error";
  error: string | null;
  response: string | null;
  steps: Step[];
  meta?: {
    total: number;
    max: number;
    questions: QuestionResult[];
    source: string;
    mode: "full" | "subset" | "general";
    instructions: string;
  };
};

const clip = (s: string, n = 600) =>
  s.length > n ? s.slice(0, n) + " …[truncated]" : s;

// A fixed sample exam graded question-by-question. Stands in for the real pipeline.
const SAMPLE_QUESTIONS: QuestionResult[] = [
  {
    id: "Q1",
    title: "Q1 · Derivative — product rule",
    score: 3,
    max: 5,
    status: "partial",
    mark: "3/5",
    feedback:
      "Product rule applied to the first term correctly. Missing the second term x²·(1/x)=x. Method sound, one term omitted → partial credit. Expected 2x·ln x + x.",
  },
  {
    id: "Q2",
    title: "Q2 · Limit — L'Hôpital",
    score: 5,
    max: 5,
    status: "ok",
    mark: "✓",
    feedback:
      "Indeterminate form 0/0 identified, L'Hôpital applied once, limit = 1. Matches the rubric solution. Full credit.",
  },
  {
    id: "Q3",
    title: "Q3 · Lagrange MVT — proof",
    score: 10,
    max: 12,
    status: "partial",
    mark: "10/12",
    feedback:
      "Auxiliary function g(x) constructed correctly and Rolle's theorem invoked. Endpoints g(a)=g(b)=0 shown. Gap: continuity/differentiability of g not stated before applying Rolle. −2.",
  },
  {
    id: "Q4",
    title: "Q4 · Integral — substitution",
    score: 7,
    max: 7,
    status: "ok",
    mark: "✓",
    feedback:
      "u-substitution u=x²+1 chosen, bounds transformed, antiderivative correct against the rubric solution. Full credit.",
  },
  {
    id: "Q5",
    title: "Q5 · Taylor series — remainder",
    score: 4,
    max: 5,
    status: "escalate",
    mark: "?",
    feedback:
      "Handwriting on the remainder bound is ambiguous (two overwritten symbols). Reader confidence low; escalated to the teacher rather than guessed.",
  },
];

function buildSteps(source: string, qs: QuestionResult[], instructions: string): Step[] {
  const steps: Step[] = [];
  const instrLine = instructions ? `\nUser instructions: ${clip(instructions, 200)}` : "";

  // 1) Reader — one OCR pass over all pages, splits the exam into questions.
  steps.push({
    module: "Reader",
    pattern: "OCR",
    prompt: {
      system_prompt:
        "Read all pages of the scanned Calculus 1 exam. Return each question's handwritten work as text/LaTeX, with a confidence score per question. Do not grade.",
      user_prompt: `[exam file: ${clip(source, 120)} — pages 1–4]${instrLine}`,
    },
    response: {
      questions: qs.map((q) => ({
        id: q.id,
        confidence: q.status === "escalate" ? 0.58 : 0.94,
      })),
    },
  });

  // Per question (graded in parallel in the real pipeline): Retriever -> Grader -> Reflector(conditional)
  for (const q of qs) {
    steps.push({
      module: "Retriever",
      pattern: "RAG",
      prompt: {
        system_prompt:
          "Match this question to the knowledge base: pull the official solution, rubric, and graded examples.",
        user_prompt: `${q.id}: ${q.title}`,
      },
      response: { matched: `rubric/${q.id}`, graded_examples: 2, max_points: q.max },
    });

    steps.push({
      module: "Grader",
      pattern: "Few-shot",
      prompt: {
        system_prompt:
          "Grade the student's OWN method against the retrieved rubric and graded examples. Award partial credit. Return score, max, and written feedback.",
        user_prompt: `Grade ${q.id} from OCR text + retrieved rubric.`,
      },
      response: { score: q.score, max: q.max, feedback: clip(q.feedback, 300) },
    });

    // Reflection is conditional — only when the grade looks shaky.
    if (q.status === "partial" || q.status === "escalate") {
      const action = q.status === "escalate" ? "ESCALATE" : "APPROVE";
      steps.push({
        module: "Reflector",
        pattern: "Reflection",
        prompt: {
          system_prompt:
            "The grade looks shaky (borderline partial credit, or low OCR confidence, or score/feedback disagree). Critique it and choose exactly ONE action: APPROVE · REVISE(≤1) · RETRY_OCR(≤1) · ESCALATE.",
          user_prompt: `Review ${q.id}: ${q.score}/${q.max}, OCR confidence ${
            q.status === "escalate" ? 0.58 : 0.94
          }.`,
        },
        response: {
          action,
          note:
            action === "ESCALATE"
              ? "OCR confidence 0.58 on the remainder bound — send to teacher."
              : "Partial credit is justified by the rubric; accept.",
        },
      });
    }
  }

  return steps;
}

// Detect "general feedback, no scoring" requests (EN + HE).
function isGeneral(lc: string): boolean {
  return /general|overall|holistic|no scor|no grad|without grad|do ?n['’]?t grade|feedback only|כללי|חוו|בלי ציון|בלי לצ|ללא ציון|בלי להתמ/.test(
    lc
  );
}

// Extract requested question numbers, e.g. "Q3", "question 4", "שאלה 3".
function requestedQuestions(instr: string): string[] {
  const ids = new Set<string>();
  const re = /(?:q|question|שאל\w*)\s*\.?\s*(\d+)/gi;
  let m: RegExpExecArray | null;
  while ((m = re.exec(instr))) ids.add("Q" + m[1]);
  return [...ids];
}

export function runMockAgent(source: string, instructions = ""): ExecuteResult {
  if (!source || !source.trim()) {
    return {
      status: "error",
      error: "No exam provided. Upload a scanned exam (PDF, Word, or image).",
      response: null,
      steps: [],
    };
  }

  const instr = (instructions || "").trim();
  const lc = instr.toLowerCase();

  // ---- General-feedback mode: holistic opinion, no per-question scoring. ----
  if (instr && isGeneral(lc)) {
    const steps: Step[] = [
      {
        module: "Reader",
        pattern: "OCR",
        prompt: {
          system_prompt:
            "Read all pages of the scanned Calculus 1 exam and return the student's work as text/LaTeX. Do not grade.",
          user_prompt: `[exam file: ${clip(source, 120)}]\nUser instructions: ${clip(instr, 200)}`,
        },
        response: { questions_read: SAMPLE_QUESTIONS.length },
      },
      {
        module: "Grader",
        pattern: "Few-shot",
        prompt: {
          system_prompt:
            "The user asked for a GENERAL opinion of the student's exam — no per-question scores. Write a short holistic assessment: overall command of the material, strengths, recurring weaknesses, and one suggestion. Do not output numeric grades.",
          user_prompt: `Give a general assessment of the exam.\nUser instructions: ${clip(instr, 200)}`,
        },
        response: {
          assessment:
            "Solid grasp of core techniques (product rule, L'Hôpital, substitution). Main weakness: stating theorem hypotheses before applying them, and completing multi-term derivatives. One answer's handwriting is unclear and would need a second look.",
        },
      },
    ];
    return {
      status: "ok",
      error: null,
      response:
        "General assessment (no scoring, as requested):\n\n" +
        "The student shows a solid command of the core Calculus 1 techniques — the product rule, L'Hôpital's rule, and u-substitution are applied correctly. " +
        "The recurring weakness is rigor around theorem hypotheses (e.g. stating continuity/differentiability before invoking Rolle's/Lagrange's theorem) and occasionally dropping a term in multi-term derivatives. " +
        "One answer's handwriting is ambiguous and would benefit from a closer read. Overall: a capable student who would gain the most from tightening proof rigor.\n\n" +
        "Note: demo build — grading is simulated.",
      steps,
      meta: { total: 0, max: 0, questions: [], source, mode: "general", instructions: instr },
    };
  }

  // ---- Full or subset grading. ----
  const wanted = requestedQuestions(instr);
  const subset = wanted.length > 0;
  const qs = subset
    ? SAMPLE_QUESTIONS.filter((q) => wanted.includes(q.id))
    : SAMPLE_QUESTIONS;

  // Requested a question we don't have in the sample exam.
  if (subset && qs.length === 0) {
    return {
      status: "error",
      error: `Requested ${wanted.join(", ")}, but the exam has ${SAMPLE_QUESTIONS.map((q) => q.id).join(", ")}.`,
      response: null,
      steps: [],
    };
  }

  const steps = buildSteps(source, qs, instr);
  const total = qs.reduce((s, q) => s + q.score, 0);
  const max = qs.reduce((s, q) => s + q.max, 0);
  const escalated = qs.filter((q) => q.status === "escalate");

  const header = subset
    ? `CheckMate graded ${qs.map((q) => q.id).join(", ")}: ${total} / ${max}.`
    : `CheckMate graded the exam: ${total} / ${max}.`;

  const lines = [
    header,
    "",
    ...qs.map((q) => `${q.id} — ${q.score}/${q.max}: ${q.feedback}`),
    "",
    escalated.length
      ? `⚠ ${escalated.length} question(s) escalated to the teacher: ${escalated
          .map((q) => q.id)
          .join(", ")}.`
      : "No questions required human review.",
    "",
    "Note: demo build — grading is simulated. The real pipeline runs OCR, Pinecone RAG, and gpt-5.4-mini grading, and returns the annotated exam.",
  ];

  return {
    status: "ok",
    error: null,
    response: lines.join("\n"),
    steps,
    meta: { total, max, questions: qs, source, mode: subset ? "subset" : "full", instructions: instr },
  };
}

File: checkmate/mock_agent.py
"""CheckMate -- mock agent. Produces a correctly-shaped /api/execute payload with a
realistic steps[] trace when no LLM key / no image is available. Port of lib/mockAgent.ts.
"""
from __future__ import annotations

import re
from dataclasses import asdict

from .models import QuestionResult


def _clip(s: str, n: int = 600) -> str:
    return s[:n] + " …[truncated]" if len(s) > n else s


# A fixed sample exam graded question-by-question. Stands in for the real pipeline.
SAMPLE_QUESTIONS: list[QuestionResult] = [
    QuestionResult("Q1", "Q1 · Derivative — product rule", 3, 5, "partial", "3/5",
        "Product rule applied to the first term correctly. Missing the second term "
        "x²·(1/x)=x. Method sound, one term omitted → partial credit. Expected 2x·ln x + x."),
    QuestionResult("Q2", "Q2 · Limit — L'Hôpital", 5, 5, "ok", "✓",
        "Indeterminate form 0/0 identified, L'Hôpital applied once, limit = 1. Matches the "
        "rubric solution. Full credit."),
    QuestionResult("Q3", "Q3 · Lagrange MVT — proof", 10, 12, "partial", "10/12",
        "Auxiliary function g(x) constructed correctly and Rolle's theorem invoked. Endpoints "
        "g(a)=g(b)=0 shown. Gap: continuity/differentiability of g not stated before applying "
        "Rolle. −2."),
    QuestionResult("Q4", "Q4 · Integral — substitution", 7, 7, "ok", "✓",
        "u-substitution u=x²+1 chosen, bounds transformed, antiderivative correct against the "
        "rubric solution. Full credit."),
    QuestionResult("Q5", "Q5 · Taylor series — remainder", 4, 5, "escalate", "?",
        "Handwriting on the remainder bound is ambiguous (two overwritten symbols). Parser "
        "confidence low; escalated to the teacher rather than guessed."),
]


def _step(module: str, pattern: str, system: str, user: str, response) -> dict:
    return {"module": module, "pattern": pattern,
            "prompt": {"System_prompt": system, "User_prompt": user}, "response": response}


def _build_steps(source: str, qs: list[QuestionResult], instructions: str) -> list[dict]:
    steps: list[dict] = []
    instr_line = f"\nUser instructions: {_clip(instructions, 200)}" if instructions else ""

    steps.append(_step(
        "Parser", "Vision OCR",
        "Read all pages of the scanned Calculus 1 exam. Return each question's handwritten "
        "work as text/LaTeX, with a confidence score per question. Do not grade.",
        f"[exam file: {_clip(source, 120)} — pages 1–4]{instr_line}",
        {"questions": [{"id": q.id, "confidence": 0.58 if q.status == "escalate" else 0.94} for q in qs]},
    ))

    for q in qs:
        steps.append(_step("Retriever", "RAG",
            "Match this question to the knowledge base: pull the official solution, rubric, "
            "and graded examples.", f"{q.id}: {q.title}",
            {"matched": f"rubric/{q.id}", "graded_examples": 2, "max_points": q.max}))
        steps.append(_step("Grader", "Few-shot",
            "Grade the student's OWN method against the retrieved rubric and graded examples. "
            "Award partial credit. Return score, max, and written feedback.",
            f"Grade {q.id} from OCR text + retrieved rubric.",
            {"score": q.score, "max": q.max, "feedback": _clip(q.feedback, 300)}))
        if q.status in ("partial", "escalate"):
            action = "ESCALATE" if q.status == "escalate" else "APPROVE"
            steps.append(_step("Reflector", "Reflection",
                "The grade looks shaky (borderline partial credit, or low OCR confidence, or "
                "score/feedback disagree). Critique it and choose exactly ONE action: "
                "APPROVE · REVISE(≤1) · RETRY_OCR(≤1) · ESCALATE.",
                f"Review {q.id}: {q.score}/{q.max}, OCR confidence {0.58 if q.status == 'escalate' else 0.94}.",
                {"action": action,
                 "note": "OCR confidence 0.58 on the remainder bound — send to teacher."
                         if action == "ESCALATE" else "Partial credit is justified by the rubric; accept."}))
    return steps


_GENERAL_RE = re.compile(
    r"general|overall|holistic|no scor|no grad|without grad|do ?n['’]?t grade|feedback only|"
    r"כללי|חוו|בלי ציון|בלי לצ|ללא ציון|בלי להתמ")


def _is_general(lc: str) -> bool:
    return bool(_GENERAL_RE.search(lc))


def _requested_questions(instr: str) -> list[str]:
    ids: list[str] = []
    for m in re.finditer(r"(?:q|question|שאל\w*)\s*\.?\s*(\d+)", instr, re.IGNORECASE):
        qid = "Q" + m.group(1)
        if qid not in ids:
            ids.append(qid)
    return ids


def run_mock_agent(source: str, instructions: str = "") -> dict:
    if not source or not source.strip():
        return {"status": "error",
                "error": "No exam provided. Upload a scanned exam (PDF, Word, or image).",
                "response": None, "steps": []}

    instr = (instructions or "").strip()
    lc = instr.lower()

    # ---- General-feedback mode: holistic opinion, no per-question scoring. ----
    if instr and _is_general(lc):
        steps = [
            _step("Parser", "Vision OCR",
                "Read all pages of the scanned Calculus 1 exam and return the student's work "
                "as text/LaTeX. Do not grade.",
                f"[exam file: {_clip(source, 120)}]\nUser instructions: {_clip(instr, 200)}",
                {"questions_read": len(SAMPLE_QUESTIONS)}),
            _step("Grader", "Few-shot",
                "The user asked for a GENERAL opinion of the student's exam — no per-question "
                "scores. Write a short holistic assessment: overall command of the material, "
                "strengths, recurring weaknesses, and one suggestion. Do not output numeric grades.",
                f"Give a general assessment of the exam.\nUser instructions: {_clip(instr, 200)}",
                {"assessment": "Solid grasp of core techniques (product rule, L'Hôpital, "
                 "substitution). Main weakness: stating theorem hypotheses before applying them, "
                 "and completing multi-term derivatives. One answer's handwriting is unclear."}),
        ]
        return {
            "status": "ok", "error": None,
            "response": (
                "General assessment (no scoring, as requested):\n\n"
                "The student shows a solid command of the core Calculus 1 techniques — the "
                "product rule, L'Hôpital's rule, and u-substitution are applied correctly. "
                "The recurring weakness is rigor around theorem hypotheses (e.g. stating "
                "continuity/differentiability before invoking Rolle's/Lagrange's theorem) and "
                "occasionally dropping a term in multi-term derivatives. One answer's handwriting "
                "is ambiguous and would benefit from a closer read. Overall: a capable student "
                "who would gain the most from tightening proof rigor.\n\n"
                "Note: demo build — grading is simulated."),
            "steps": steps,
            "meta": {"total": 0, "max": 0, "questions": [], "source": source,
                     "mode": "general", "instructions": instr},
        }

    # ---- Full or subset grading. ----
    wanted = _requested_questions(instr)
    subset = len(wanted) > 0
    qs = [q for q in SAMPLE_QUESTIONS if q.id in wanted] if subset else list(SAMPLE_QUESTIONS)

    if subset and not qs:
        have = ", ".join(q.id for q in SAMPLE_QUESTIONS)
        return {"status": "error", "error": f"Requested {', '.join(wanted)}, but the exam has {have}.",
                "response": None, "steps": []}

    steps = _build_steps(source, qs, instr)
    total = sum(q.score for q in qs)
    max_pts = sum(q.max for q in qs)
    escalated = [q for q in qs if q.status == "escalate"]

    header = (f"CheckMate graded {', '.join(q.id for q in qs)}: {total} / {max_pts}."
              if subset else f"CheckMate graded the exam: {total} / {max_pts}.")
    lines = [
        header, "",
        *[f"{q.id} — {q.score}/{q.max}: {q.feedback}" for q in qs], "",
        (f"⚠ {len(escalated)} question(s) escalated to the teacher: "
         f"{', '.join(q.id for q in escalated)}.") if escalated else "No questions required human review.",
        "",
        "Note: demo build — grading is simulated. The real pipeline runs OCR, Pinecone RAG, "
        "and gpt-5.4-mini grading, and returns the annotated exam.",
    ]

    return {
        "status": "ok", "error": None, "response": "\n".join(lines), "steps": steps,
        "meta": {"total": total, "max": max_pts, "questions": [asdict(q) for q in qs],
                 "source": source, "mode": "subset" if subset else "full", "instructions": instr},
    }
